server/grpc/rpc.go

```diff
@@ -429,6 +488,43 @@ func (s *RPC) ReportHealth(ctx context.Context, status string) error {
 	return s.store.AgentUpdate(agent)
 }
 
+func (s *RPC) checkAgentPermissionByWorkflow(_ context.Context, agent *model.Agent, strWorkflowID string, pipeline *model.Pipeline, repo *model.Repo) error {
+	var err error
+	if repo == nil {
+		if pipeline == nil {
+			workflowID, err := strconv.ParseInt(strWorkflowID, 10, 64)
+			if err != nil {
+				return err
+			}
+
+			workflow, err := s.store.WorkflowLoad(workflowID)
+			if err != nil {
+				log.Error().Err(err).Msgf("rpc.update: cannot find workflow with id %d", workflowID)
+				return err
+			}
+
+			pipeline, err = s.store.GetPipeline(workflow.PipelineID)
+			if err != nil {
+				log.Error().Err(err).Msgf("cannot find pipeline with id %d", workflow.PipelineID)
+				return err
+			}
+		}
+
+		repo, err = s.store.GetRepo(pipeline.RepoID)
+		if err != nil {
+			log.Error().Err(err).Msgf("cannot find repo with id %d", pipeline.RepoID)
+			return err
+		}
+	}
+
+	if !agent.CanAccessRepo(repo) {
+		msg := fmt.Sprintf("agent '%d' is not allowed to interact with repo[%d] '%s'", agent.ID, repo.ID, repo.FullName)
+		log.Error().Int64("repoId", repo.ID).Msg(msg)
+		return errors.New(msg)
+	}
+	return nil
```

Comment:
For a security check it would be better to fail by default and `return nil` if everything is fine.

Reply:
fd6a92e07c577ee084d2fb4d8391e426b2444bab